[Merge] 用户登录功能实现的合并请求

FLASHLACK1314 opened the pull request:
用户登录功能实现请求实现，已经经过接口测试，缺少单元测试

xiaolfeng committed:
refactor(scheduling): 为 UserLoginDTO 中的字段添加 @NotNull 注解

- 在 UserLoginDTO 类中，为 user、token 和 initialization 字段添加了 @NotNull 注解
- 这些更改可以增强数据传输对象的健壮性，确保在使用这些字段时不会出现空指针异常

Commit:
feat(user): 添加用户登录功能- 新增 StudentDAO、TeacherDAO 和 TokenDAO
- 创建 UserLogic 实现用户登录逻辑
- 在 UserController 中集成用户登录功能
- 添加 UserService 接口和 UserLoginDTO、UserDTO 等数据传输对象

Commit:
Merge remote-tracking branch 'origin/develop' into feature/flashlack/user-login

# Conflicts:
#	src/main/java/com/frontleaves/scheduling/daos/TokenDAO.java

xiaolfeng committed:
docs(README): 添加项目 README 文档

- 新增 README.md 文件，介绍智课方舟项目的背景、目标和主要功能
- 概述了智能排课、手工排课调整、实时冲突检测、多维数据统计分析、课表管理与导出、智能调课等功能模块
- 为项目提供了详细的说明文档，有助于用户和开发者了解项目的基本信息和使用方法

xiaolfeng committed:
feat(TokenDAO): 增强 Token 管理功能

- 添加了 `ErrorCode` 和 `BusinessException` 导入。
- 更新了类注释，详细描述了 Token 的创建、验证、刷新和删除操作。
- 在 `createToken` 方法中增加了详细的注释，并明确了参数和返回值的说明。
- 在 `verifyToken` 方法中增加了异常处理，当 Token 不属于当前用户时抛出 `BusinessException`。
- 新增了 `refreshToken` 方法，用于刷新 Token 的过期时间，并在 RefreshToken 过期时更新 Redis 中的数据。
- 新增了 `deleteToken` 方法，用于删除指定的 Token，并在 Token 不属于当前用户时抛出异常。
- 在 `StringConstant` 类中新增了 `TOKEN_ATTRIBUTION_ERROR` 常量。

Commit:
feat(login): 完善用户登录逻辑

- 新增通过学号获取学生信息和通过工号获取教师信息的方法
- 修改登录逻辑，支持学生和教师的初始化登录和正常登录
- 优化密码错误提示，统一为"用户名或密码错误"
- 修复未初始化用户登录后返回信息不完整的问题

Commit:
feat(user): 新增用户注册和登录功能

- 新增 MajorDAO、MajorDO 和 MajorMapper 类- 新增 UserInitializationVO 类用于用户初始化
- 修改 UserController 中的 userRegistered 方法，增加用户注册逻辑
- 更新 UserLogic 类，增加用户登录时生成 token 的逻辑- 新增 UserTest 类用于测试用户登录功能

xiaolfeng committed:
fix(models): 修改UserLoginDTO中user和token的注解为Nullable

- 将UserLoginDTO中的user字段从@NotNull修改为@Nullable
- 将UserLoginDTO中的token字段从@NotNull修改为@Nullable

xiaolfeng committed:
fix(models): 修改UserLoginDTO中user和token的注解为Nullable

- 将UserLoginDTO中的user字段从@NotNull修改为@Nullable
- 将UserLoginDTO中的token字段从@NotNull修改为@Nullable

Commit:
feat(user): 添加用户注册功能并优化登录逻辑

- 在 UserController 中添加用户注册方法
- 在 UserInitializationVO 中增加密码、邮箱和手机号字段，并添加验证注解- 在 UserLogic 中实现用户注册逻辑，包括信息校验和用户创建
- 在 UserService 中添加 userRegistered 方法接口
- 删除了 UserTest 类

Commit:
refactor: 移除用户相关接口的多余注释

- 删除了 UserController 类中用户登录和用户注册接口的多余注释
- 优化了代码结构，提高了代码的可读性和维护性

Commit:
refactor(user): 优化用户注册和登录逻辑

- 修改了用户注册和登录的接口和逻辑实现
- 增加了密码验证和加密功能
- 优化了数据库表名和字段名- 调整了实体类和 VO 类的结构

Commit:
fix(user): 修复用户初始化和登录相关问题

- 在 UserInitializationVO 中添加了新密码的非空校验
- 修改了用户登录时的密码验证逻辑，使用 PasswordUtil.verify 方法
- 修复了用户初始化时密码加密的问题，使用 newPassword 替代 password
- 更新了用户角色名称，将"教师"改为"老师"

Commit:
refactor(scheduling): 优化用户登录验证逻辑

- 将原有的串行查询改为根据输入类型（邮箱、手机号、用户名）进行针对性查询
- 使用正则表达式匹配不同类型的用户标识，提高验证准确性
- 优化日志信息，明确记录登录方式

Commit:
refactor(UserLogic): 移除初始化检测注释

移除了 UserLogic 类中不必要的注释，该注释用于检测用户是否已初始化。这个改动简化了代码结构，提高了代码的可读性。

Commit:
refactor(scheduling): 移除多余的换行符

- 在 UserLogic 类中的初始化用户逻辑中删除了一个多余的换行符
-优化了代码格式，提高了代码的可读性

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/logic/UserLogic.java:
```diff
@@ -0,0 +1,280 @@
+                userLoginDTO = loginReturn(userDO);
+                return userLoginDTO;
+            } else {
+                throw new BusinessException("用户名或密码错误", ErrorCode.BODY_ERROR);
```
重复部分写入 constants 里，对于字符串重复部分 StringConstant

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/logic/UserLogic.java:
```diff
@@ -0,0 +1,280 @@
+                throw new BusinessException("用户名或密码错误", ErrorCode.BODY_ERROR);
+            }
+        } else {
+            log.info("确认为是否为学号登录，检查是否初始化");
```
日志有效内容用 info，开发部分用 debug

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/logic/UserLogic.java:
```diff
@@ -0,0 +1,280 @@
+                log.info("确认为学号登录，检查是否初始化");
+                if (studentDO.getUserUuid() != null) {
+                    log.info("学号登录已经初始化过");
+                    UserDO userDoS = userDAO.getById(studentDO.getUserUuid());
```
重复代码处理

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/logic/UserLogic.java:
```diff
@@ -0,0 +1,280 @@
+                            //正常登录
+                            return loginReturn(userDoS);
+                        } else {
+                            throw new BusinessException("用户名或密码错误", ErrorCode.BODY_ERROR);
```
存在一个具体关于授权相关报错的，UserAuthenticationException

xiaolfeng commented on src/main/java/com/frontleaves/scheduling/models/vo/UserInitializationVO.java:
```diff
@@ -0,0 +1,36 @@
+            message = "强密码(必须包含大小写字母和数字的组合，可以使用特殊字符，至少6位）")
+    @NotBlank(message = "新密码不能为空")
+    private String newPassword;
+    @Pattern(regexp = "^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,6}$",
```
大部分正则表达式这里有：https://www.jyshare.com/front-end/854/

xiaolfeng reviewed (changes requested):
重复部分提取，代码逻辑该优化了。下一个插件 SonarQube 关于代码规范的

xiaolfeng committed:
fix(filters): 更新InitialModeFilter和RequestHeaderFilter

- 更新InitialModeFilter的文档注释，明确其在初始化模式下的作用。
- 在RequestHeaderFilter中添加日志记录，增强对请求头检查失败时的警告信息。

xiaolfeng committed:
feat(init_mode): 修正初始化模式检查逻辑并添加系统初始化状态检查接口

- 修改 `InitialModeFilter` 中的 URI 检查方式，使用 `startsWith` 替换 `contains` 以提高准确性。
- 在 `InitController` 中添加了新的 `systemInitCheck` 方法，用于检查系统是否处于初始化模式，并返回相应的状态信息。
- 引入了 `SystemInitCheckDTO` 类，用于封装系统初始化状态的响应数据。

xiaolfeng committed:
refactor(constant): 更新常量命名并增强 TokenDTO 字段校验

- 将 `StringConstant.Common` 中的常量替换为 `StringConstant.Common.Hump` 中的驼峰命名常量。
- 在 `TokenDTO` 类中添加 `@NotNull` 注解以增强字段校验。
- 调整 `TokenDAO` 中的相关方法，使用新的驼峰命名常量。

Commit:
refactor(authentication): 重构用户登录验证逻辑

- 将正则表达式移至 StringConstant 类统一管理
- 优化用户登录验证逻辑，提高代码可读性和维护性
- 引入 UserAuthenticationException 异常处理登录错误
- 通过 HttpServletRequest 对象获取用户请求信息

xiaolfeng committed:
docs(copyright): 添加版权声明和许可证声明

- 在多个文件的头部添加了版权声明和MIT许可证声明。
- 更新了部分正则表达式的常量名称，以符合新的命名规范。
- 增加了类和方法的注释，增强了代码的可读性和维护性。

Commit:
feat(user): 支持学生和教师用户登录

- 新增学号和工号登录功能
- 实现用户初始化注册和密码检查
- 优化登录逻辑，支持多种登录方式
- 新增用户信息获取方法

xiaolfeng committed:
Merge branch 'develop' into feature/flashlack/user-login

# Conflicts:
#	src/main/java/com/frontleaves/scheduling/constants/StringConstant.java

Commit:
refactor(scheduling): 重构字符串常量类并更新正则表达式使用

- 重构 StringConstant 类，将正则表达式常量移动到 Common 内部类中
- 更新 UserInitializationVO 和 UserLogic 中的正则表达式引用
- 删除 StringConstant 类中未使用的 TOKEN_ATTRIBUTION_ERROR常量
- 优化 StringConstant 类的注释和结构

xiaolfeng committed:
refactor(UserLogic,UserDAO): 优化用户登录逻辑并引入日志常量

- 在 `UserLogic` 中添加了 `@NotNull` 注解以确保 `UserLoginVO` 不为空。
- 更新了 `UserLogic` 中的日志级别和消息格式，使用新的 `LogConstant` 类统一日志输出。
- 在 `UserDAO` 中添加了 Redis 缓存支持，以提高查询性能。
- 引入了 `LogConstant` 类，用于统一和标准化日志输出的前缀。

xiaolfeng committed:
fix(pattern): 更新正则表达式常量位置并修正部分注释

- 将`UserInitializationVO.java`中使用的正则表达式常量从`StringConstant.Common.Hump`移动到`StringConstant.Regular`
- 更新`StringConstant.java`中的注释，使其更加清晰和准确
- 修正`StringConstant`类中部分字符串常量的命名，如将`USER_TEL`改为小写`user:tel:`

Commit:
refactor(scheduling): 优化用户登录和注册逻辑

- 重新组织用户登录和注册相关代码，提高可读性和可维护性

Commit:
Merge branch 'master' into feature/flashlack/user-login

FLASHLACK1314 commented:
@XiaoLFeng 请求合并♪（＾∀＾●）ﾉｼ （●´∀｀）♪

xiaolfeng committed:
Merge branch 'develop' into feature/flashlack/user-login

# Conflicts:
#	src/main/java/com/frontleaves/scheduling/daos/UserDAO.java
#	src/main/java/com/frontleaves/scheduling/logic/UserLogic.java

xiaolfeng committed:
refactor(UserLogic,PermissionDAO,InitController,RoleDAO): 重构用户逻辑和服务接口

- 重构 `UserLogic` 类中的方法，优化了用户登录和注册的逻辑处理。
- 在 `PermissionDAO` 中增加了异常处理，确保在 Redis 或数据库操作失败时抛出 `ServerInternalErrorException`。
- 修改了 `InitController` 中获取管理员角色的方式，从 `RoleDO` 改为 `RoleDTO`。
- 更新了 `RoleDAO` 的依赖注入，引入了 Hutool 工具类以简化对象转换。

xiaolfeng committed:
feat(nosql): 添加新用户登录检查方法并重构相关逻辑

- 新增 `checkLoginForNewUserByStudent` 和 `checkLoginForNewUserByTeacher` 方法，用于处理新用户通过学生或教师信息登录。
- 重构 `checkLoginForNewUser` 方法，调用上述新增方法以简化代码。
- 在 `buildLoginData` 方法中添加对用户权限为空时的处理。
- 修改 `StudentDAO` 和 `TeacherDAO` 中的部分方法，确保返回正确的对象。
